eks/deploy_state.go

```diff
@@ -0,0 +1,360 @@
+	asgInfo, err := getAsgInfo(asgSvc, eksAsgName)
+	if err != nil {
+		return err
+	}
+
+	// Calculate default max retries
+	state.maxRetries = ensureMaxRetries(state.maxRetries, state.sleepBetweenRetries, asgInfo.OriginalCapacity)
+
+	// Make sure ASG is in steady state
+	if asgInfo.OriginalCapacity != int64(len(asgInfo.OriginalInstances)) {
+		state.logger.Infof("Ensuring ASG is in steady state (current capacity = desired capacity)")
+		err = waitForCapacity(asgSvc, eksAsgName, state.maxRetries, state.sleepBetweenRetries)
+		if err != nil {
+			state.logger.Error("Error waiting for ASG to reach steady state. Try again after the ASG is in a steady state.")
+			return err
+		}
+		state.logger.Infof("Verified ASG is in steady state (current capacity = desired capacity)")
+		asgInfo, err = getAsgInfo(asgSvc, eksAsgName)
+		if err != nil {
+			return err
+		}
+	}
+
+	state.GatherASGInfoDone = true
+	state.ASGs = append(state.ASGs, asgInfo)
+	return state.persist()
+}
+
+// ensureMaxRetries ensures we always have a proper value for maxRetries, either set by the end user
+// or calculated based on the original capacity
+func ensureMaxRetries(maxRetries int, sleepBetweenRetries time.Duration, originalCapacity int64) int {
+	logger := logging.GetProjectLogger()
+	// Calculate default max retries
+	if maxRetries == 0 {
+		defaultMaxRetries := getDefaultMaxRetries(originalCapacity, sleepBetweenRetries)
+		logger.Infof(
+			"No max retries set. Defaulted to %d based on sleep between retries duration of %s and scale up count %d.",
+			defaultMaxRetries,
+			sleepBetweenRetries,
+			originalCapacity,
+		)
+		return defaultMaxRetries
+	} else {
+		return maxRetries
+	}
+}
+
+// setMaxCapacity will set the max size of the auto scaling group.
+func (state *DeployState) setMaxCapacity(asgSvc *autoscaling.AutoScaling) error {
+	if state.SetMaxCapacityDone {
+		state.logger.Debug("Max capacity already set - skipping")
+		return nil
+	}
+	asg := &state.ASGs[0]
+	maxCapacityForUpdate := asg.OriginalCapacity * 2
+	if asg.OriginalMaxCapacity < maxCapacityForUpdate {
+		err := setAsgMaxSize(asgSvc, asg.Name, maxCapacityForUpdate)
+		if err != nil {
+			return err
+		}
+	}
+	asg.MaxCapacityForUpdate = maxCapacityForUpdate
+	state.SetMaxCapacityDone = true
+	return state.persist()
+}
+
+// scaleUp will scale up the ASG and wait until all the nodes are available.
+func (state *DeployState) scaleUp(asgSvc *autoscaling.AutoScaling) error {
+	if state.ScaleUpDone {
+		state.logger.Debug("Scale up already done - skipping")
+		return nil
+	}
+	asg := &state.ASGs[0]
+	state.logger.Info("Starting with the following list of instances in ASG:")
+	state.logger.Infof("%s", strings.Join(asg.OriginalInstances, ","))
+
+	state.logger.Infof("Launching new nodes with new launch config on ASG %s", asg.Name)
+	newInstanceIds, err := scaleUp(asgSvc, asg.Name, asg.OriginalInstances, asg.MaxCapacityForUpdate, state.maxRetries, state.sleepBetweenRetries)
```

Comment:
I just noticed this. How does Go know which `scaleUp` is being referred to here, is it via the parameters (the function signature)? We have a DeployState scaleUp and an ASG scaleUp.

Reply:
The `scaleUp` function for `DeployState` is a method and not a function. E.g., it would call the `DeployState` version (and recurse) if this was `state.scaleUp`.

Reply:
Ah, right! Thank you.